diff --git a/Makefile b/Makefile
index b262c0a..e3b1392 100644
--- a/Makefile
+++ b/Makefile
@@ -145,6 +145,7 @@ UPROGS=\
 	$U/_logstress\
 	$U/_forphan\
 	$U/_dorphan\
+	$U/_test_scheduler\
 
 fs.img: mkfs/mkfs README $(UPROGS)
 	mkfs/mkfs fs.img README $(UPROGS)
@@ -166,7 +167,7 @@ QEMUGDB = $(shell if $(QEMU) -help | grep -q '^-gdb'; \
 	then echo "-gdb tcp::$(GDBPORT)"; \
 	else echo "-s -p $(GDBPORT)"; fi)
 ifndef CPUS
-CPUS := 3
+CPUS := 1
 endif
 
 QEMUOPTS = -machine virt -bios none -kernel $K/kernel -m 128M -smp $(CPUS) -nographic
diff --git a/kernel/proc.c b/kernel/proc.c
index 22a5401..9a008fc 100644
--- a/kernel/proc.c
+++ b/kernel/proc.c
@@ -6,6 +6,8 @@
 #include "proc.h"
 #include "defs.h"
 
+int rand(void); 
+
 struct cpu cpus[NCPU];
 
 struct proc proc[NPROC];
@@ -125,6 +127,9 @@ found:
   p->pid = allocpid();
   p->state = USED;
 
+  p->tickets = 1;  
+  p->rounds = 0;
+
   // Allocate a trapframe page.
   if((p->trapframe = (struct trapframe *)kalloc()) == 0){
     freeproc(p);
@@ -228,6 +233,8 @@ userinit(void)
 
   p->state = RUNNABLE;
 
+  p->tickets = 10; //for initial process to have 10 tickets
+
   release(&p->lock);
 }
 
@@ -279,6 +286,9 @@ kfork(void)
   // copy saved user registers.
   *(np->trapframe) = *(p->trapframe);
 
+  np->tickets = p->tickets;
+  np->rounds = 0;
+
   // Cause fork to return 0 in the child.
   np->trapframe->a0 = 0;
 
@@ -421,47 +431,66 @@ kwait(uint64 addr)
 //  - swtch to start running that process.
 //  - eventually that process transfers control
 //    via swtch back to the scheduler.
+
+
 void
 scheduler(void)
 {
-  struct proc *p;
   struct cpu *c = mycpu();
-
   c->proc = 0;
+
   for(;;){
-    // The most recent process to run may have had interrupts
-    // turned off; enable them to avoid a deadlock if all
-    // processes are waiting. Then turn them back off
-    // to avoid a possible race between an interrupt
-    // and wfi.
     intr_on();
-    intr_off();
 
-    int found = 0;
-    for(p = proc; p < &proc[NPROC]; p++) {
+    int total_tickets = 0;
+
+    // Step 1: count total tickets
+    for(struct proc *p = proc; p < &proc[NPROC]; p++){
       acquire(&p->lock);
-      if(p->state == RUNNABLE) {
-        // Switch to chosen process.  It is the process's job
-        // to release its lock and then reacquire it
-        // before jumping back to us.
-        p->state = RUNNING;
-        c->proc = p;
-        swtch(&c->context, &p->context);
-
-        // Process is done running for now.
-        // It should have changed its p->state before coming back.
-        c->proc = 0;
-        found = 1;
-      }
+      if(p->state == RUNNABLE)
+        total_tickets += p->tickets;
       release(&p->lock);
     }
-    if(found == 0) {
-      // nothing to run; stop running on this core until an interrupt.
+
+    if(total_tickets == 0){
       asm volatile("wfi");
+      continue;
+    }
+
+    // Step 2: pick winning ticket
+    int winning = rand() % total_tickets;
+
+    // Step 3: find winner
+    int current = 0;
+    struct proc *winner = 0;
+
+    for(struct proc *p = proc; p < &proc[NPROC]; p++){
+      acquire(&p->lock);
+      if(p->state == RUNNABLE){
+        current += p->tickets;
+        if(current > winning){
+          winner = p;
+          break;
+        }
+      }
+      release(&p->lock);
+    }
+
+    // Step 4: run winner
+    if(winner){
+      winner->rounds++;
+      winner->state = RUNNING;
+      c->proc = winner;
+
+      swtch(&c->context, &winner->context);
+
+      c->proc = 0;
+      release(&winner->lock);
     }
   }
 }
 
+
 // Switch to scheduler.  Must hold only p->lock
 // and have changed proc->state. Saves and restores
 // intena because intena is a property of this
@@ -684,7 +713,32 @@ procdump(void)
       state = states[p->state];
     else
       state = "???";
-    printf("%d %s %s", p->pid, state, p->name);
+    printf("%d %s %s tickets=%d rounds=%d\n",p->pid, state, p->name, p->tickets, p->rounds);
     printf("\n");
   }
 }
+
+
+unsigned long rand_next = 1;
+
+int
+do_rand(unsigned long *ctx)
+{
+  long hi, lo, x;
+
+  x = (*ctx % 0x7ffffffe) + 1;
+  hi = x / 127773;
+  lo = x % 127773;
+  x = 16807 * lo - 2836 * hi;
+  if (x < 0)
+    x += 0x7fffffff;
+  x--;
+  *ctx = x;
+  return x;
+}
+
+int
+rand(void)
+{
+  return do_rand(&rand_next);
+}
\ No newline at end of file
diff --git a/kernel/proc.h b/kernel/proc.h
index d021857..10cd4a8 100644
--- a/kernel/proc.h
+++ b/kernel/proc.h
@@ -91,7 +91,10 @@ struct proc {
   int killed;                  // If non-zero, have been killed
   int xstate;                  // Exit status to be returned to parent's wait
   int pid;                     // Process ID
-
+  
+  int tickets;   // number of lottery tickets
+  int rounds;    // number of times scheduled
+  
   // wait_lock must be held when using this:
   struct proc *parent;         // Parent process
 
diff --git a/kernel/syscall.c b/kernel/syscall.c
index 076d965..ca57370 100644
--- a/kernel/syscall.c
+++ b/kernel/syscall.c
@@ -80,6 +80,8 @@ argstr(int n, char *buf, int max)
 }
 
 // Prototypes for the functions that handle system calls.
+extern uint64 sys_settickets(void);
+
 extern uint64 sys_fork(void);
 extern uint64 sys_exit(void);
 extern uint64 sys_wait(void);
@@ -126,6 +128,8 @@ static uint64 (*syscalls[])(void) = {
 [SYS_link]    sys_link,
 [SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
+
+[SYS_settickets] sys_settickets,
 };
 
 void
diff --git a/kernel/syscall.h b/kernel/syscall.h
index 3dd926d..f085330 100644
--- a/kernel/syscall.h
+++ b/kernel/syscall.h
@@ -20,3 +20,4 @@
 #define SYS_link   19
 #define SYS_mkdir  20
 #define SYS_close  21
+#define SYS_settickets 22
\ No newline at end of file
diff --git a/kernel/sysproc.c b/kernel/sysproc.c
index 419e727..846dd09 100644
--- a/kernel/sysproc.c
+++ b/kernel/sysproc.c
@@ -7,6 +7,23 @@
 #include "proc.h"
 #include "vm.h"
 
+uint64
+sys_settickets(void)
+{
+  int n;
+  argint(0, &n);
+
+  if(n <= 0)
+    return -1;
+
+  struct proc *p = myproc();
+  acquire(&p->lock);
+  p->tickets = n;
+  release(&p->lock);
+
+  return 0;
+}
+
 uint64
 sys_exit(void)
 {
diff --git a/user/addDemo.c b/user/addDemo.c
new file mode 100644
index 0000000..617b484
--- /dev/null
+++ b/user/addDemo.c
@@ -0,0 +1,10 @@
+// user/addDemo.c
+#include "kernel/types.h"
+#include "user/user.h"
+
+int main(){
+int add = 2+3;
+printf("add: %d\n",add);
+
+return 0;
+}
\ No newline at end of file
diff --git a/user/add_with_arg.c b/user/add_with_arg.c
new file mode 100644
index 0000000..f00a1f1
--- /dev/null
+++ b/user/add_with_arg.c
@@ -0,0 +1,22 @@
+// user/add_with_arg.c
+
+#include "kernel/types.h"
+#include "user/user.h"
+
+int main(int argc, char *argv[]){
+    if(argc < 3){
+    printf("Addition failed_____No arguments given_____\n");
+    }
+
+    else if(argc == 3){
+        int a = atoi(argv[1]);
+        int b = atoi(argv[2]);
+        int add = a + b;
+        printf("addition of 2 inputs %d and %d = %d\n",a,b,add);
+        }
+    else{
+        printf("Addition failed_____Too many arguments given_____\n");
+    }
+    
+    return 0;
+}
\ No newline at end of file
diff --git a/user/ps.c b/user/ps.c
new file mode 100644
index 0000000..bcd0def
--- /dev/null
+++ b/user/ps.c
@@ -0,0 +1,9 @@
+// user/ps.c
+#include "kernel/types.h"
+#include "user/user.h"
+int
+main(void)
+{
+ps(); // call the ps system call
+exit(0);
+}
\ No newline at end of file
diff --git a/user/test_priority.c b/user/test_priority.c
new file mode 100644
index 0000000..17a54bc
--- /dev/null
+++ b/user/test_priority.c
@@ -0,0 +1,21 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int
+main(int argc, char *argv[])
+{
+  if(argc < 2){
+    printf("Usage: test_priority <priority>\n");
+    exit(1);
+  }
+
+  int pr = atoi(argv[1]);
+  setpriority(pr);
+
+  while(1) {
+    // busy loop
+  }
+
+  exit(0);
+}
diff --git a/user/test_scheduler.c b/user/test_scheduler.c
new file mode 100644
index 0000000..c3bcc6b
--- /dev/null
+++ b/user/test_scheduler.c
@@ -0,0 +1,19 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int
+main(int argc, char *argv[])
+{
+  if(argc < 2){
+    printf("Usage: test_scheduler <tickets>\n");
+    exit(1);
+  }
+
+  int t = atoi(argv[1]);
+  settickets(t);
+
+  while(1) { }
+
+  exit(0);
+}
\ No newline at end of file
diff --git a/user/user.h b/user/user.h
index ac84de9..e00c487 100644
--- a/user/user.h
+++ b/user/user.h
@@ -25,6 +25,8 @@ char* sys_sbrk(int,int);
 int pause(int);
 int uptime(void);
 
+int settickets(int);
+
 // ulib.c
 int stat(const char*, struct stat*);
 char* strcpy(char*, const char*);
diff --git a/user/usys.pl b/user/usys.pl
index c5d4c3a..d12241d 100755
--- a/user/usys.pl
+++ b/user/usys.pl
@@ -42,3 +42,4 @@ entry("getpid");
 entry("sbrk");
 entry("pause");
 entry("uptime");
+entry("settickets");ss
\ No newline at end of file
